common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -300,59 +337,106 @@ void unguardedToAvailable() {
         }
     }
 
-    private static final class LocalPool<T> implements MessagePassingQueue.Consumer<DefaultHandle<T>> {
+    private static final class GuardedLocalPool<T> extends LocalPool<DefaultHandle<T>, T> {
+
+        GuardedLocalPool(Thread owner, int maxCapacity, int ratioInterval, int chunkSize) {
+            super(owner, maxCapacity, ratioInterval, chunkSize);
+        }
+
+        GuardedLocalPool(int maxCapacity, int ratioInterval, int chunkSize) {
+            super(maxCapacity, ratioInterval, chunkSize);
+        }
+
+        @Override
+        public T getWith(Recycler<T> recycler) {
+            DefaultHandle<T> handle = acquire();
+            T obj;
+            if (handle == null) {
+                handle = canAllocatePooled()? new DefaultHandle<>(this) : null;
+                if (handle != null) {
+                    obj = recycler.newObject(handle);
+                    handle.set(obj);
+                } else {
+                    obj = recycler.newObject((Handle<T>) NOOP_HANDLE);
+                }
+            } else {
+                obj = handle.claim();
+            }
+            return obj;
+        }
+    }
+
+    private static final class UnguardedLocalPool<T> extends LocalPool<T, T> {
+        private final EnhancedHandle<T> handle;
+
+        UnguardedLocalPool(Thread owner, int maxCapacity, int ratioInterval, int chunkSize) {
+            super(owner, maxCapacity, ratioInterval, chunkSize);
+            handle = new LocalPoolHandle<>(this);
+        }
+
+        UnguardedLocalPool(int maxCapacity, int ratioInterval, int chunkSize) {
+            super(maxCapacity, ratioInterval, chunkSize);
+            handle = new LocalPoolHandle<>(this);
+        }
+
+        @Override
+        public T getWith(Recycler<T> recycler) {
+            T obj = acquire();
+            if (obj == null) {
+                obj = recycler.newObject(canAllocatePooled()? handle : (Handle<T>) NOOP_HANDLE);
+            }
+            return obj;
+        }
+    }
+
+    private abstract static class LocalPool<H, T> {
         private final int ratioInterval;
         private final int chunkSize;
-        private final ArrayDeque<DefaultHandle<T>> batch;
+        private final ArrayList<H> batch;
         private volatile Thread owner;
```

Comment:
I don't see a value for this - the fast thread local can just use `pooledHandles` to store release this